Remove more unused elements

Squareys opened the pull request:
Hi everybody!

While writing a custom stage for gltf-pipeline, I had the need to remove unused materials, meshes and nodes.

I figured this might not always be desirable, which is why I added `--removeUnused material,node,accessor` to `gtlf-pipeline.js` to allow specification of which types of elements should be removed.
To keep backwards compatibility, this defaults to `accessor,buffer,bufferView`.

Since order of removal matters, I ensured that the order is always the same independent of the order of given element types.

Are you interested in this pullrequest? Then I will make sure to test this and clean it up, as this is currently just some code extracted from my custom stage.

Looking forward to your reply,
Jonathan.

lilleyse commented:
Hi @Squareys,

Yes we're interested in this. Just one thing, I think we should always default to removing all unused elements and have a `--keep-unused-elements` option to disable that behavior. I'm not sure if we need the level of granularity currently in the PR, at least not yet.

Then the question is what to do about accessors, bufferViews, and buffers that need to be cleaned up from internal stages like `compressDracoMeshes` when `--keep-unused-elements` is passed in. I think each stage needs to be more careful about cleaning up after itself. I could open a PR for just that.

lilleyse commented:
Could you also fill out a CLA so we can review this?

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla

Squareys commented:
@lilleyse Done! Sorry I missed that :)

No need to review _yet_, I wanted to roughly probe interest for merging this kind of work first! I should do some cleanup in case there is.

Squareys commented:
> --keep-unused-elements

I can imagine a usecase where you'd want to keep "empty" nodes (as weapon sockets or snapping something to in games for example), but still want to remove meshes and buffers. 

Similarly, you may want to keep materials, e.g. some gltf for a configurator application that contains the mesh and different unreferenced materials to switch at runtime. But still want to optimize the rest of the scene 🤔

pjcozzi commented:
Thanks again for the contribution @Squareys, we received your CLA.

Squareys commented:
@lilleyse In case you missed it, I had answered with some concerns whether the granularity may be useful after all. Wanted to make sure we come to a decision before cleaning this up.

lilleyse commented:
@Squareys Those are good use cases but I'm not sure if it happens much in practice where the glTF has a mix of intentionally unused properties and unintentionally unused properties. I still think we should keep the API simple to start.

Squareys commented:
@lilleyse Internally, I kept the granularity to be able to provide backwards compatible behaviour of draco mesh compression, update version, split primitives and a couple of other pipeline stages. (Otherwise most test cases would break, as usually there are materials, meshes and other elements that are not referenced and would be removed.)

I added the `--keepUnusedObjects` which will keep unused nodes, materials and meshes, restoring the behaviour of the current version.

I will look into the CI failure ASAP (maybe later today), I think that's the linting.

lilleyse commented:
@Squareys Thanks for the update, all sounds good.

Squareys commented:
@lilleyse This is ready for review!

I fixed the code style issues and improved test coverage a bit, added to README and CHANGES and finally rewrote the git commit history.

Squareys commented:
Rebased onto latest master

lilleyse commented:
Thanks, I hope to review this soon.

Squareys commented:
Hey @lilleyse, I don't want to be obtrusive, just wanted to make sure you push this PR back into my responsibility in case there is any need to clean up more. If any of the changes are unclear, I'm happy to answer any questions :)

lilleyse commented:
Sorry for the delay, it's still on my radar though. I might have some time later in the week to start the review.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -260,3 +324,81 @@ getListOfElementsIdsInUse.bufferView = function(gltf) {
+}
+
+getListOfElementsIdsInUse.node = function(gltf) {
+    const usedNodeIds = {};
```
Also handle:

* `skin.skeleton` and `skin.joints` referencing nodes
* [`KHR_techniques_webgl`](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_techniques_webgl) referencing a node. Which would look like:

```
    ForEach.technique(gltf, function (technique) {
        ForEach.techniqueUniform(technique, function (uniform) {
            if (defined(uniform.node)) {
                usedNodeIds[uniform.node] = true;
            }
        });
    });
```

Squareys replied:
Awesome! Was expecting something like this to pop up, thanks!

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -12,20 +12,28 @@ module.exports = removeUnusedElements;
  * This function currently only works for accessors, buffers, and bufferViews.
```
Some overall comments for this file:

* Avoid any features above es5 because certain files of gltf-pipeline including this one are bundled in CesiumJS which still supports IE11. It's an annoying limitation that isn't made very clear to new contributors, so sorry about that. `const` and `let` are fine though and will work once https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/471 is merged. A summary of the bundling situation is here: https://github.com/AnalyticalGraphicsInc/gltf-pipeline/issues/330.
* For style consistency, change `if(` to `if (` everywhere.
* Also style related, use `//` for single line comments rather than `/*  */`

Squareys replied:
> bundled in CesiumJS which still supports IE11

Ah! I was very curious about why you don't use some of the newer features. I suspect in the tests they are ok, cause they are not bundled, right?

lilleyse replied:
Yeah, exactly.

Squareys replied:
Alright, I changed:

* `if(` to `if (`
* `x => {}` to `function(x) {}`
* `/* */` to `//`
* default parameter of removeUnusedElements

I hope the two were all the ES6 functions I used (Went through [this resource](http://es6-features.org/) to find them)

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -158,6 +166,62 @@ Remove.bufferView = function(gltf, bufferViewId) {
+
+    nodes.splice(nodeId, 1);
+
+    /* Shift all node references */
```
Also shift node references for skeletons, joints, animations, and `KHR_technique_webgl`.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -158,6 +166,62 @@ Remove.bufferView = function(gltf, bufferViewId) {
+
+    /* Shift other material ids */
+    ForEach.mesh(gltf, (m, j) => {
+        m.primitives.forEach(p => {
```
Use `ForEach.meshPrimitive`

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -158,6 +166,62 @@ Remove.bufferView = function(gltf, bufferViewId) {
+    nodes.splice(nodeId, 1);
+
+    /* Shift all node references */
+    nodes.forEach(other => {
```
Use `ForEach.node`.

lilleyse commented on lib/removeUnusedElements.js:
```diff
@@ -158,6 +166,62 @@ Remove.bufferView = function(gltf, bufferViewId) {
+    ForEach.scene(gltf, (s, i) => {
+        s.nodes = s.nodes
+            .filter(x => x !== nodeId) /* Remove */
+            .map(x => x > nodeId ? x - 1 : x); /* Shift indices */
```
Could this same filter/map code be used above to combine the remove-from-parent and shift-indices steps? Maybe something like:

```js
    const nodes = gltf.nodes;
    nodes.splice(nodeId, 1);

    ForEach.node(gltf, (n) => {
        n.children = n.children
            .filter(x => x !== nodeId) /* Remove */
            .map(x => x > nodeId ? x - 1 : x); /* Shift indices */
    });
```

Squareys commented:
Oh! Sorry, this time the delay was on my side...

Will get to this over the weekend, will comment some replies in case I have any questions there 👍 
Thanks for the review!

Squareys commented:
@lilleyse Done! Once you're happy with the changes, I can rebase to auto squash in those `fixup!` commits.

Tell me if there is anything I missed or anything else you want me to change :)

lilleyse commented:
@Squareys I pushed some fixes and did a consistency/cleanup pass over the code. For the sake of not holding this PR up any longer I'm going to go ahead and merge.

Thanks again!
